<a href="https://colab.research.google.com/github/almansmaa/py/blob/main/warmup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem 1. A fair game should have an expected value of 0. There are 6 ways to roll a 7 out of 36 possible rolls, so the probability of winning is \(6/36=1/6\). The expected value is \(4(1/6)-1(5/6)=-1/6\). Since the expected value is negative, this game is not fair. The winning payout would need to be 5 chips instead of 4 for the expected value to equal 0.

In [ ]:
import random

times = 1000000
money = 0

for i in range(times):
    d1 = random.randint(1, 6)
    d2 = random.randint(1, 6)
    sum = d1 + d2
    if sum == 7:
        money += 4
    else:
        money -= 1

print(money/times)

-0.16524


The result is close to -0.167 which means that the player. loses about 0.167 chips per game on average. This is close to -1/6 so the game is unfair.

In [ ]:
count = 0

for d1 in range(1, 7):
    for d2 in range(1, 7):
        for d3 in range(1, 7):
            sum = d1 + d2 + d3
            if sum not in [3, 5, 7, 11, 13, 17]:
                count += 1
            else:
                pass

print(count/(6 * 6 * 6))

0.6620370370370371


The result is approximately 0.662 which means that there is. a 66.2 chance that the sum of the three dice is not a prime number. The simulation is close to the exact answer so it makes sense.


In [ ]:
lockers = [False] * 100

for student in range(1, 101):
    for locker in range(student - 1, 100, student):
        lockers[locker] = not lockers[locker]

    open_lockers = []
    for i in range(100):
        if lockers[i]:
            open_lockers.append(i + 1)

    print("After student", student, "open lockers:", open_lockers)

print("Lockers open at the end:")

for i in range(100):
    if lockers[i]:
        print(i + 1, end=" ")

After student 1 open lockers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]
After student 2 open lockers: [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99]
After student 3 open lockers: [1, 5, 6, 7, 11, 12, 13, 17, 18, 19, 23, 24, 25, 29, 30, 31, 35, 36, 37, 41, 42, 43, 47, 48, 49, 53, 54, 55, 59, 60, 61, 65, 66, 67, 71, 72, 73, 77, 78, 79, 83, 84, 85, 89, 90, 91, 95, 96, 97]
After student 4 open lockers: [1, 4, 5, 6, 7, 8, 11, 13, 16, 17, 18, 19, 20, 23, 25, 28, 29, 30, 31, 32, 35, 37, 40, 41, 42, 43,

The lockers that are open at the end are 1, 4, 9, 16, 25, 36, 49, 64, 81, and 100. These are all perfect squares.